File: ZoomiesServer/app/src/main/java/com/example/zoomies_server/MainActivity.java
package com.example.zoomies_server;

import android.net.wifi.WifiInfo;
import android.net.wifi.WifiManager;
import android.os.Bundle;
import android.text.method.ScrollingMovementMethod;
import android.widget.EditText;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;

import com.example.zoomies_server.model.database.AppDatabase;
import com.example.zoomies_server.database.dto.AnimalDTO;
import com.example.zoomies_server.database.dto.GeneralDTO;
import com.example.zoomies_server.model.dto.UserDTO;
import com.example.zoomies_server.model.database.entity.UserRole;
import com.example.zoomies_server.model.dto.factory.AnimalDTOFactory;
import com.example.zoomies_server.model.dto.factory.UserDTOFactory;
import com.example.zoomies_server.viewmodel.MainViewModel;
import com.google.gson.Gson;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.net.UnknownHostException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    ServerSocket serverSocket;
    Thread Thread1 = null;
    TextView tvIP, tvPort;
    TextView tvMessages;
    EditText etMessage;
    public static String SERVER_IP = "";
    public static final int SERVER_PORT = 8080;
    String message;

    public static AppDatabase database;
    private MainViewModel viewModel;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        database = AppDatabase.Companion.getDatabase(this);
        viewModel = new ViewModelProvider(this).get(MainViewModel.class);
        viewModel.initContext(this);

        tvIP = findViewById(R.id.tvIP);
        tvPort = findViewById(R.id.tvPort);
        tvMessages = findViewById(R.id.tvMessages);
        tvMessages.setMovementMethod(new ScrollingMovementMethod());
        etMessage = findViewById(R.id.etMessage);
        try {
            SERVER_IP = getLocalIpAddress();
        } catch (UnknownHostException e) {
            e.printStackTrace();
        }
        Thread1 = new Thread(new Connector());
        Thread1.start();
    }

    private String getLocalIpAddress() throws UnknownHostException {
        WifiManager wifiManager = (WifiManager) getApplicationContext().getSystemService(WIFI_SERVICE);
        assert wifiManager != null;
        WifiInfo wifiInfo = wifiManager.getConnectionInfo();
        int ipInt = wifiInfo.getIpAddress();
        return InetAddress.getByAddress(ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(ipInt).array()).getHostAddress();
    }

    private PrintWriter output;
    private BufferedReader input;

    class Connector implements Runnable {
        @Override
        public void run() {
            Socket socket;
            try {
                serverSocket = new ServerSocket(SERVER_PORT);
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        tvMessages.setText("Not connected");
                        tvIP.setText("IP: " + SERVER_IP);
                        tvPort.setText("Port: " + String.valueOf(SERVER_PORT));
                    }
                });
                try {
                    socket = serverSocket.accept();
                    output = new PrintWriter(socket.getOutputStream());
                    input = new BufferedReader(new InputStreamReader(socket.getInputStream()));
                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            tvMessages.setText("Connected" + "\n" + "\n");
                        }
                    });

                    new Thread(new Sender(AppDatabase.Companion.getAnimals(), AppDatabase.Companion.getUsers())).start();

                    new Thread(new Receiver()).start();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    class Receiver implements Runnable {
        @Override
        public void run() {
            while (true) {
                try {
                    String message = input.readLine();
                    GeneralDTO transfer = (new Gson().fromJson(message, GeneralDTO.class));
                    String messageToDisplay = "";
                    if (transfer.getUserNameS() != null) {
                        UserDTO userDTO = UserDTOFactory.Companion.getInstance().createDTO(null, "", "", "", "", transfer.getUidS(), transfer.getUserNameS(), transfer.getPasswordS(), transfer.getRoleS(), transfer.getEmailS(), transfer.getPhoneNumberS());
                        if (transfer.getRequestDelete()) {
                            viewModel.deleteUser(userDTO);
                            messageToDisplay = "delete user:\n" + userDTO + "\n" + "\n";
                        } else {
                            viewModel.insertUser(userDTO);
                            messageToDisplay = "insert user:\n" + userDTO + "\n" + "\n";
                        }
                    } else {
                        AnimalDTO animalDTO = AnimalDTOFactory.Companion.getInstance().createDTO(transfer.getAnimalIdS(), transfer.getNameS(), transfer.getSpeciesS(), transfer.getHabitatS(), transfer.getDietS(), null, "", "", UserRole.EMPLOYEE, "", "");
                        if (transfer.getRequestDelete()) {
                            viewModel.deleteAnimal(animalDTO);
                            messageToDisplay = "delete animal:\n" + animalDTO + "\n" + "\n";
                        } else {
                            viewModel.insertAnimal(animalDTO);
                            messageToDisplay = "insert animal:\n" + animalDTO + "\n" + "\n";
                        }
                    }

                    if (message != null) {
                        String finalMessageToDisplay = messageToDisplay;
                        runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                tvMessages.append("received package:" + finalMessageToDisplay + "\n" + "\n");
                            }
                        });
                    } else {
                        Thread1 = new Thread(new Connector());
                        Thread1.start();
                        return;
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    class Sender implements Runnable {
        List<AnimalDTO> animalDTOList;
        List<UserDTO> userDTOList;

        public Sender(List<AnimalDTO> animalDTOList, List<UserDTO> userDTOList) {
            this.animalDTOList = animalDTOList;
            this.userDTOList = userDTOList;
        }

        @Override
        public void run() {
            List<GeneralDTO> usersGeneralDTOS = new ArrayList<>(userDTOList);
            String serializedUsers = new Gson().toJson(usersGeneralDTOS);
            output.println(serializedUsers);

            List<GeneralDTO> animalsGeneralDTOS;
            if (animalDTOList == null) {
                animalsGeneralDTOS = new ArrayList<>();
            } else {
                animalsGeneralDTOS = new ArrayList<>(animalDTOList);
            }

            String serializedAnimals = new Gson().toJson(animalsGeneralDTOS);
            output.println(serializedAnimals);
            output.flush();

            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    tvMessages.append("sent package:" + "animals:\n" + animalDTOList + "\n");
                    tvMessages.append("sent package:" + "users:\n" + userDTOList + "\n");
                }
            });
        }
    }
}

File: Zoomies2/app/src/main/java/com/example/zoomies/MainActivity.kt
package com.example.zoomies

import android.content.Intent
import android.net.Uri
import android.net.wifi.WifiManager
import android.os.Build
import android.os.Bundle
import android.util.Log
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.annotation.RequiresApi
import androidx.lifecycle.MutableLiveData
import com.example.zoomies.model.AppSettingsProvider
import com.example.zoomies.model.dto.AnimalDTO
import com.example.zoomies.model.dto.SerializationUtils
import com.example.zoomies.model.dto.GeneralDTO
import com.example.zoomies.model.dto.UserDTO
import com.example.zoomies.model.dto.factory.AnimalDTOFactory
import com.example.zoomies.model.dto.factory.UserDTOFactory
import com.example.zoomies.view.components.AppNavHost
import com.example.zoomies.view.theme.ZoomiesTheme
import com.google.gson.Gson
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.io.PrintWriter
import java.net.InetAddress
import java.net.Socket
import java.net.UnknownHostException
import java.nio.ByteBuffer
import java.nio.ByteOrder

class MainActivity : ComponentActivity() {

    @RequiresApi(Build.VERSION_CODES.R)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        AppSettingsProvider.initializeSharedPreferences(this)
        getLocalIpAddress()?.let {
            SERVER_IP = it
        }
        setContent {
            ZoomiesTheme {
                AppNavHost()
            }
        }

    }

    fun sendEmail(email: String, body: String) {
        val subject = "Zoomies Account"
        val addresses: Array<String> = listOf(email).toTypedArray()

        val emailIntent = Intent(Intent.ACTION_SENDTO).apply {
            data = Uri.parse("mailto:")
            putExtra(Intent.EXTRA_EMAIL, addresses)
            putExtra(Intent.EXTRA_SUBJECT, subject)
            putExtra(Intent.EXTRA_TEXT, body)
        }
        startActivity(Intent.createChooser(emailIntent, "Send Email"), null)
    }

    fun sendSMS(phoneNumber: String, message: String) {
        val uri = Uri.parse("smsto:$phoneNumber")
        val intent = Intent(Intent.ACTION_SENDTO, uri)
        intent.putExtra("sms_body", message)
        startActivity(intent)
    }

    @Throws(UnknownHostException::class)
    private fun getLocalIpAddress(): String? {
        val wifiManager = (applicationContext.getSystemService(WIFI_SERVICE) as WifiManager)
        val wifiInfo = wifiManager.connectionInfo
        val ipInt = wifiInfo.ipAddress
        return InetAddress.getByAddress(
            ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(ipInt).array()
        ).hostAddress
    }

    internal class Connector : Runnable {
        override fun run() {
            val socket: Socket
            try {
                socket = Socket(SERVER_IP, SERVER_PORT)
                output = PrintWriter(socket.getOutputStream())
                input = BufferedReader(InputStreamReader(socket.getInputStream()))

                Thread(Receiver()).start()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

    internal class Receiver : Runnable {
        override fun run() {
            while (true) {
                try {
                    val message = input!!.readLine()
                    if (message != null) {
                        val transfer: List<GeneralDTO> = SerializationUtils.castTransfer(message)
                        if (transfer.isNotEmpty() && transfer.first().animalIdS != null) {
                            val animalList: MutableList<AnimalDTO> = mutableListOf()
                            transfer.forEach {
                                animalList.add(
                                    AnimalDTOFactory.instance.createDTO(
                                        animalId = it.animalIdS,
                                        name = it.nameS!!,
                                        species = it.speciesS!!,
                                        habitat = it.habitatS!!,
                                        diet = it.dietS!!
                                    )
                                )
                            }
                            animals = animalList
                            serverRequestFinished.postValue(Unit)
                        } else {
                            if (transfer.isNotEmpty() && transfer.first().userNameS != null) {
                                val userList: MutableList<UserDTO> = mutableListOf()
                                transfer.forEach {
                                    userList.add(
                                        UserDTOFactory.instance.createDTO(
                                            uid = it.uidS,
                                            userName = it.userNameS!!,
                                            password = it.passwordS!!,
                                            role = it.roleS!!,
                                            email = it.emailS!!,
                                            phoneNumber = it.phoneNumberS!!
                                        )
                                    )
                                }
                                users = userList
                                serverRequestFinished.postValue(Unit)
                            }
                        }
                    } else {
                        connectorThread = Thread(Connector())
                        connectorThread.start()
                        return
                    }
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        }
    }

    internal class Sender(private val generalDTO: GeneralDTO) : Runnable {
        override fun run() {
            output!!.println(Gson().toJson(generalDTO))
            output!!.flush()
        }
    }


    fun attemptConnectionToServer(ip: String, port: Int = 8080) {
        SERVER_IP = ip
        SERVER_PORT = port
        connectorThread = Thread(Connector());
        connectorThread.start();
    }

    fun requestInsertAnimal(animalDTO: AnimalDTO) {
        val generalDTO: GeneralDTO = animalDTO
        generalDTO.requestInsert = true
        Thread(Sender(generalDTO)).start()
    }

    fun requestInsertUser(userDTO: UserDTO) {
        val generalDTO: GeneralDTO = userDTO
        generalDTO.requestInsert = true
        Thread(Sender(generalDTO)).start()
    }

    fun requestDeleteAnimal(animalDTO: AnimalDTO) {
        val generalDTO: GeneralDTO = animalDTO
        generalDTO.requestDelete = true
        Thread(Sender(generalDTO)).start()
    }

    fun requestDeleteUser(userDTO: UserDTO) {
        val generalDTO: GeneralDTO = userDTO
        generalDTO.requestDelete = true
        Thread(Sender(generalDTO)).start()
    }

    companion object {
        var SERVER_IP: String = ""
        var SERVER_PORT = 8080

        var output: PrintWriter? = null
        var input: BufferedReader? = null

        var animals: List<AnimalDTO>? = null
        var users: List<UserDTO>? = null

        val serverRequestFinished = MutableLiveData<Unit>()

        private lateinit var connectorThread: Thread
    }
}